src/contrib/cluster/Akka.Cluster.Sharding.Tests.MultiNode/AsyncWriteProxyEx.cs

```diff
@@ -0,0 +1,461 @@
+//     Copyright (C) 2013-2016 Akka.NET project <https://github.com/akkadotnet/akka.net>
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Akka.Actor;
+using System.Runtime.Serialization;
+using Akka.Event;
+using Akka.Persistence.Journal;
+using Akka.Persistence;
+using System.Threading;
+using Akka.Util.Internal;
+using Akka.Actor.Internal;
+
+namespace Akka.Cluster.Sharding.Tests
+{
+    /// <summary>
+    /// This exception is thrown when the replay inactivity exceeds a specified timeout.
+    /// </summary>
+    [Serializable]
+    public class AsyncReplayTimeoutException : AkkaException
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AsyncReplayTimeoutException"/> class.
+        /// </summary>
+        public AsyncReplayTimeoutException()
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AsyncReplayTimeoutException"/> class.
+        /// </summary>
+        /// <param name="message">The message that describes the error.</param>
+        public AsyncReplayTimeoutException(string message)
+            : base(message)
+        {
+        }
+
+#if SERIALIZATION
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AsyncReplayTimeoutException"/> class.
+        /// </summary>
+        /// <param name="info">The <see cref="SerializationInfo"/> that holds the serialized object data about the exception being thrown.</param>
+        /// <param name="context">The <see cref="StreamingContext"/> that contains contextual information about the source or destination.</param>
+        protected AsyncReplayTimeoutException(SerializationInfo info, StreamingContext context)
+            : base(info, context)
+        {
+        }
+#endif
+    }
+
+    /// <summary>
+    /// TBD
+    /// </summary>
+    [Serializable]
+    public sealed class SetStore
+    {
+        /// <summary>
+        /// TBD
+        /// </summary>
+        /// <param name="store">TBD</param>
+        /// <exception cref="ArgumentNullException">
+        /// This exception is thrown when the specified <paramref name="store"/> is undefined.
+        /// </exception>
+        public SetStore(IActorRef store)
+        {
+            if (store == null)
+                throw new ArgumentNullException(nameof(store), "SetStore requires non-null reference to store actor");
+
+            Store = store;
+        }
+
+        /// <summary>
+        /// TBD
+        /// </summary>
+        public readonly IActorRef Store;
+    }
+
+    /// <summary>
+    /// A journal that delegates actual storage to a target actor. For testing only.
+    /// </summary>
+    public abstract class AsyncWriteProxyEx : AsyncWriteJournal, IWithUnboundedStash
```

Comment:
Enhanced version of AsyncWriteProxy, since the one in Persistence didn't work out.
It's simulating AsyncWriteJournal with sending messages to the read actor via Ask with timeout handling